### PR TITLE
upgrade permissions to 128-bit with ability to scale indefinitely

### LINKED ISSUE #1969: Permissions in future library versions
Currently, DSharpPlus models permissions as bitfield enum with 64 bits. Discord documents permissions as unbounded-length integers, and as we are approaching the 64-bit limit (at the time of writing, we are at 50 permissions), we will need to change how we handle permissions.

Permissions will be refactored into a `DiscordPermission` enum containing all individual permissions, and a `DiscordPermissions`/`DiscordPermissionSet` struct storing all these permissions. Additionally, serialization will be changed to tolerate permissions being too long, so as to ensure Discord rolling over any future threshold will not render all previous builds of the library unusable.

## Proposed API Surface

```diff
-[Flags]
-public enum DiscordPermissions;
+public enum DiscordPermission;
```

```cs
public struct DiscordPermissions
{
    public DiscordPermissions(DiscordPermission);
    public DiscordPermissions(params ReadOnlySpan<DiscordPermission>);
    public DiscordPermissions(BigInteger);
    public DiscordPermissions(scoped ReadOnlySpan<byte>);

    public static implicit operator DiscordPermissions(DiscordPermission);

    public bool HasFlag(DiscordPermission);
    public bool HasPermission(DiscordPermission); // factors in admin
    public bool HasAnyPermission(params ReadOnlySpan<DiscordPermission>);
    public bool HasAllPermissions(params ReadOnlySpan<DiscordPermission>);
    public DiscordPermissions Toggle(DiscordPermission); // toggles the value of the field on or off
    public DiscordPermissions Toggle(params ReadOnlySpan<DiscordPermission>); // toggles the value of the fields on or off
    
    public static bool operator ==(DiscordPermissions, DiscordPermissions);
    public static bool operator !=(DiscordPermissions, DiscordPermissions);
    
    public DiscordPermissions Add(DiscordPermission);
    public DiscordPermissions Add(params ReadOnlySpan<DiscordPermission>);
    public static DiscordPermissions operator +(DiscordPermissions, DiscordPermission);
    
    public DiscordPermissions Remove(DiscordPermission);
    public DiscordPermissions Remove(params ReadOnlySpan<DiscordPermission>);
    public static DiscordPermissions operator -(DiscordPermissions, DiscordPermission);
    public static DiscordPermissions operator -(DiscordPermissions, DiscordPermissions);
    
    public static DiscordPermissions operator &(DiscordPermissions, DiscordPermissions);
    public static DiscordPermissions operator &(DiscordPermissions, DiscordPermission);
    public static DiscordPermissions operator |(DiscordPermissions, DiscordPermissions);
    public static DiscordPermissions operator |(DiscordPermissions, DiscordPermission);
    public static DiscordPermissions operator ^(DiscordPermissions, DiscordPermissions);
    public static DiscordPermissions operator ^(DiscordPermissions, DiscordPermission);
    public static DiscordPermissions operator ~(DiscordPermissions);
}
```

Let us know on the issue or on Discord if this doesn't meet your expectation, and if you wish to discuss the details of the backing implementation, our discussion channel is found [here](https://canary.discord.com/channels/379378609942560770/1253073283323007088).

## Plan of Action

- [x] Relax v4 serialization to permit longer permissions. This does not imply changing the permission model in v4 or changing how many permissions v4 can represent.
- [x] Implement the new design in v5 and v6 (v6 can serve as experimenting ground and to explore using newer runtime/language features v5 cannot yet use)
- [x] Migrate all existing API surface interacting with permissions
  - [x] This especially concerns `DiscordChannel.PermissionsFor(DiscordMember)` and `DiscordMember.PermissionsIn(DiscordChannel)`, which work with overwrites. We might want to consider designing these entirely separately.

### PR DESCRIPTION
**Pull request by akiraveliara**
resolves #1969 
requires .NET 9 - do not merge before .NET 9 is released

this pull request enables permissions to scale indefinitely by the change of a single constant (at 512 bits we should consider using 256-bit steps and Vector256 instead, but that'll be a problem for in several decades assuming Discord lives that long without ever changing the fundamentals of their permission system). the permission logic is efficient and well-tested. the naming for the container type may still change - do not consider the API laid out here final.

the minimum size for new permissions is 128-bit for simplicity, not because it's required yet, but it'll be required soon enough regardless, possibly even before a v5 release.

### CODE REVIEW

File: DSharpPlus/Entities/DiscordPermissions.Enumeration.cs.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    /// Represents an enumerator for permission fields within a permission set.
+    /// </summary>
+    public struct DiscordPermissionEnumerator : IEnumerator<DiscordPermission>
+    {
+        private readonly DiscordPermissionContainer data;
+
+        internal DiscordPermissionEnumerator(DiscordPermissionContainer data)
+            => this.data = data;
+
+        public readonly DiscordPermission Current => (DiscordPermission)(this.block << 5) + this.bit;
```
Comment (VelvetToroyashi) [inline DSharpPlus/Entities/DiscordPermissions.Enumeration.cs.cs:44]: Dare I ask the thought process behind this math

**Review by VelvetToroyashi — APPROVED** [review]
GitHub makes a lot of this unfathomably abhorrent to actually look at, but I'm fond of the changes. I don't see anything major at first-glance, but I'm also partially sleep deprived.

File: DSharpPlus/Entities/DiscordPermissions.Enumeration.cs.cs (hunk 1 above)
Comment (akiraveliara) [reply DSharpPlus/Entities/DiscordPermissions.Enumeration.cs.cs:44]: each block is 32 bits, so `this.block * 32 + this.bit`

**Review by Plerx2493 — APPROVED** [review]
Looks good, although the bit math is a bit too high for me after work :when:

**Pull request merged**